### PR TITLE
Provide util class for reducer callback

### PR DESCRIPTION
**Pull request by drupol**
This PR

* [x] Provide a `ReducerBuilder` functional and stateless class
* [x] Replace the use of custom reducer builder by using the new `ReducerBuilder` class. 
* [x] Has tests

### CODE REVIEW

File: src/Operation/Reject.php [hunk 1]
```diff
@@ -80,11 +55,7 @@ static function (Iterator $iterator) use ($callbacks): Generator {
                         $callbacks;
 
                     foreach ($iterator as $key => $current) {
-                        $result = array_reduce(
-                            $callbacks,
-                            $reducerCallback($key)($current)($iterator),
-                            false
-                        );
+                        $result = OrCallbackReducer::or()($callbacks, $current, $key, $iterator);
```
Comment (AlexandruGG) [inline src/Operation/Reject.php:58]: I think having `Or` twice in this is a bit redundant.

I propose either:

```suggestion
                        $result = CallbackReducer::or()($callbacks, $current, $key, $iterator);
```

this leaves room to have `CallbackReducer::and()`

OR 

```suggestion
                        $result = OrCallbackReducer::of()($callbacks, $current, $key, $iterator);
```

and we make it invokable

Comment (drupol) [reply src/Operation/Reject.php]: Fair enough, done !

File: spec/loophp/collection/Utils/CallbacksArrayReducerSpec.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        $this::or()($callbacks, 0, 0, $iterator)
+            ->shouldReturn(true);
+
+        $this::or()($callbacks, 13, 0, $iterator)
+            ->shouldReturn(true);
+
+        $this::or()($callbacks, 3, 0, $iterator)
+            ->shouldReturn(false);
+    }
```
Comment (AlexandruGG) [inline spec/loophp/collection/Utils/CallbacksArrayReducerSpec.php:42]: A few comments here:
- the `$iterator` is not really used in this test because none of the callbacks use it. Would be good to have one test which uses it
- similarly, would be good to have a test with callbacks that use the second argument (the key)
- we should also test some edge cases, e.g. empty callbacks array or a single callback
- it's good practice to have one assertion per test. I know with `CollectionSpec` maybe that's not too convenient because it's a big class but with something like this I think it makes it easier to follow, e.g.
`it_reduces_empty_callbacks_array`
`it_reduces_single_callback`
`it_reduces_callbacks_with_key_and_value`
`it_reduces_callbacks_with_all_parameters`

Comment (drupol) [reply spec/loophp/collection/Utils/CallbacksArrayReducerSpec.php:42]: Ok, will add this tonight !

Comment (drupol) [reply spec/loophp/collection/Utils/CallbacksArrayReducerSpec.php]: Job done!

**Review by AlexandruGG — APPROVED** [review]
Nice one 🎉

File: src/Utils/CallbacksArrayReducer.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * @immutable
+ *
+ * @template TKey
+ * @template T
+ *
+ * phpcs:disable Generic.Files.LineLength.TooLong
+ */
+final class CallbacksArrayReducer
```
Comment (AlexandruGG) [inline src/Utils/CallbacksArrayReducer.php:23]: We could shorten the name to `CallbacksReducer`. Due to the plural it's quite clear we are dealing with a list of multiple callbacks

**Comment by drupol** [comment]
Thanks :+1:

File: src/Utils/CallbacksArrayReducer.php (hunk 3 above)
Comment (drupol) [reply src/Utils/CallbacksArrayReducer.php:23]: Oops, too late, we can tackle that in a follow-up PR...
